File: leetcode/move_zeroes.cpp
#include <iostream>
#include <vector>

using namespace std;

void moveZeroes(vector<int>& nums) {
    size_t l = 0;
    size_t r = 0;

    while (r < nums.size()) {
        if (nums[r] == 0) {
            ++r;
            continue;
        }

        nums[l] = nums[r];
        ++l;
        ++r;
    }

    for (size_t i = l; i < nums.size(); ++i) {
        nums[i] = 0;
    }
}

int main() {
    vector<int> v{0,1,0,3,12};

    moveZeroes(v);

    for (const auto & i : v) {
        cout << i << ' ';
    }
    cout << endl;

    return 0;
}

File: leetcode/MoveZeroes.java
import java.util.*;

class MoveZeroes {
    public static void moveZeroes(int[] nums) {
        int idx = 0;

        for (int n : nums) {
            if (n != 0) {
                nums[idx] = n;
                ++idx;
            }
        }

        for (int i = idx; i < nums.length; ++i) {
            nums[i] = 0;
        }
    }

    public static void main(String[] args) {
        int[] nums = new int[]{1, 0, 1};
        moveZeroes(nums);
        System.out.println(Arrays.toString(nums));
    }
}
